### PR TITLE
v5: NextJS improvements

### PR DESCRIPTION
**Pull request by nikosdouvlis**
## Description
The `auth().redirectToSignIn()` helper no longer needs to be explicitly returned when called within the middleware. The following examples are now equivalent:

```js
// Before
export default clerkMiddleware(auth => {
  if (protectedRoute && !auth.user) {
    return auth().redirectToSignIn()
  }
})

// After
export default clerkMiddleware(auth => {
  if (protectedRoute && !auth.user) {
    auth().redirectToSignIn()
  }
})
```

Calling `auth().protect()` from a page will now automatically redirect back to the same page by setting `redirect_url` to the request url before the redirect to the sign-in URL takes place.

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/backend`
- [ ] `@clerk/chrome-extension`
- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `@clerk/localizations`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/remix`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/themes`
- [ ] `@clerk/types`
- [ ] `build/tooling/chore`

### CODE REVIEW

File: packages/nextjs/src/server/clerkMiddleware.ts [hunk 1]
```diff
@@ -121,6 +108,9 @@ export const clerkMiddleware: ClerkMiddleware = (...args: unknown[]): any => {
     }
 
     decorateRequest(clerkRequest, handlerResult, requestState);
+
+    // TODO @nikos: we need to make this more generic
+    // and move the logic in clerk/backend
```
Comment (panteliselef) [inline packages/nextjs/src/server/clerkMiddleware.ts:113]: Are you referring to a possible helper named `copyHeaders` ?

**Review by panteliselef — COMMENTED** [review]
This looks good to me, Is this still WIP ?

**Comment by nikosdouvlis** [comment]
> This looks good to me, Is this still WIP ?

No :) please review it!

File: packages/backend/src/createRedirect.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { errorThrower, parsePublishableKey } from './util/shared';
 
-const buildUrl = (_baseUrl: string | URL, _targetUrl: string | URL, _returnBackUrl?: string | URL) => {
+const buildUrl = (_baseUrl: string | URL, _targetUrl: string | URL, _returnBackUrl?: string | URL | null) => {
```
Comment (dimkl) [inline packages/backend/src/createRedirect.ts:3]: ❓  why did we prefix all the params with `_`? I thought we prefix a variable with an underscore to mark it as ignored which is not the case here. Should we rename the params in another PR?

File: packages/nextjs/src/app-router/server/auth.ts [hunk 3]
```diff
@@ -19,16 +20,23 @@ export const auth = (): Auth => {
...
-  const redirectToSignIn = createRedirect({
-    redirectAdapter: redirect,
-    baseUrl: createClerkRequest(request).clerkUrl.toString(),
-    // TODO: Support runtime-value configuration of these options
-    // via setting and reading headers from clerkMiddleware
-    publishableKey: PUBLISHABLE_KEY,
-    signInUrl: SIGN_IN_URL,
-    signUpUrl: SIGN_UP_URL,
-  }).redirectToSignIn;
+  const clerkUrl = getAuthKeyFromRequest(request, 'ClerkUrl') as AuthStatus;
```
Comment (dimkl) [inline packages/nextjs/src/app-router/server/auth.ts:23]: ☁️ `clerkUrl` is of type `AuthStatus` ? Is this some kind of typo? Also `getAuthKeyFromRequest` returns `string | null | undefined`, so the `clerkUrl` many not have a `.toString()` method.

Comment (nikosdouvlis) [reply packages/nextjs/src/app-router/server/auth.ts:23]: Good catch @dimkl 
TS did not complain as AuthStatus overlaps with string. Let me fix it!

**Review by anagstef — APPROVED** [review]
💯

**Pull request merged**